ambry-quota/src/main/java/com/github/ambry/quota/MySqlStorageUsageRefresher.java

```diff
@@ -0,0 +1,470 @@
+        // Error already been logged from backup file manager.
+      }
+    }
+  }
+
+  /**
+   * Fetch the container monthly storage usage from mysql and update the in memory cache. If somehow it fails, we will
+   * retry for several times before giving up.
+   */
+  private void fetchMonthlyStorageUsageAndMaybeRetry() {
+    boolean shouldRetry = false;
+    try {
+      String monthValue = accountStatsMySqlStore.queryRecordedMonth(clusterMapConfig.clusterMapClusterName);
+      if (monthValue.equals(currentMonth)) {
+        Map<String, Map<String, Long>> base =
+            accountStatsMySqlStore.queryMonthlyAggregatedStats(clusterMapConfig.clusterMapClusterName);
+        containerStorageUsageMonthlyBase = base;
+      } else {
+        shouldRetry = true;
+      }
+    } catch (Exception e) {
+      logger.error("Failed to refresh monthly storage usage", e);
+      shouldRetry = true;
+    }
+
+    if (shouldRetry) {
+      if (retries >= config.mysqlStoreRetryMax) {
+        logger.error("Failed to refresh monthly storage usage after " + retries + " retries, will skip for month: "
+            + currentMonth);
+        retries = 0;
+      }
+      retries++;
+      if (config.mysqlStoreRetryBackoffMs != 0) {
+        scheduler.schedule(this::fetchMonthlyStorageUsageAndMaybeRetry, config.mysqlStoreRetryBackoffMs,
+            TimeUnit.MILLISECONDS);
+      }
+      return;
+    } else {
+      retries = 0;
+      tryPersistMonthlyUsage();
+      scheduleStorageUsageMonthlyBaseFetcher();
+    }
+  }
+
+  /**
+   * Schedule the task to fetch storage usage monthly base for next month.
+   */
+  private void scheduleStorageUsageMonthlyBaseFetcher() {
+    long sleepDurationInSecs = secondsToNextTick(currentMonth, config.mysqlMonthlyBaseFetchOffsetSec);
+    logger.info("Schedule to fetch container storage monthly base after " + sleepDurationInSecs + " seconds");
+    scheduler.schedule(this::fetchStorageUsageMonthlyBase, sleepDurationInSecs, TimeUnit.SECONDS);
+  }
+
+  /**
+   * The task to fetch storage usage monthly base.
+   */
+  void fetchStorageUsageMonthlyBase() {
+    currentMonth = getCurrentMonth();
+    fetchMonthlyStorageUsageAndMaybeRetry();
+  }
+
+  /**
+   * Return {@link #containerStorageUsageMonthlyBase}. Only used in test.
+   * @return the {@link #containerStorageUsageMonthlyBase}.
+   */
+  Map<String, Map<String, Long>> getContainerStorageUsageMonthlyBase() {
+    return Collections.unmodifiableMap(containerStorageUsageMonthlyBase);
+  }
+
+  /**
+   * Return {@link BackupFileManager}. Only used in test.
+   * @return {@link BackupFileManager}.
+   */
+  BackupFileManager getBackupFileManager() {
+    return backupFileManager;
+  }
+
+  /**
+   * Fetch the storage usage and schedule a task to periodically refresh the storage usage.
+   */
+  private void initialFetchAndSchedule() {
+    Runnable updater = () -> {
+      try {
+        Map<String, Map<String, Long>> base = containerStorageUsageMonthlyBase;
```

Comment:
In bootstrap case or `containerStorageUsageMonthlyBase` cannot be fetched from Mysql or local backup, shouldn't we give it an empty map at line 85?

Reply:
It's kinda dangerous to do so. If initialing monthly base failed and we give it an empty map, and somehow fetching total storage usage succeeded, then the container usage for this month would highly likely to surpass the quota. For instance, if a container's total usage is 100GB and the monthly base is 95 GB and the quota is 10 GB. If everything works fine, then for this month, it already used 100 - 95 = 5GB, the quota is 10GB, it still has 5GB to use. But if we give monthly base an empty map, then for this month, we will consider this container using 100GB already.

This is a tricky one, I suppose we don't want to stop the process just because we can't get the monthly base for quota service, but we still want to have a correct logic going forward. That's why I come to this compromise. Monthly base would be highly likely fetched from backup files but we still fail the process if we can't initialize it from database. However, we don't fail the process if we can't get it later.

Reply:
I see your point, but in the edge case aforementioned, is it possible that `containerStorageUsageMonthlyBase` is not initialized?  I have a minor concern the "base" might be null and there is no check before passing it to `subtract` method where NPE may occur.